graphql/schemabuilder/helpers.go

```diff
@@ -0,0 +1,504 @@
+	o.PaginatedFields = append(o.PaginatedFields,
+		PaginationObject{
+			Name:     name,
+			NodeType: reflect.TypeOf(obj),
+			Fun:      f,
+		})
+}
+
+func (sb *schemaBuilder) GetPaginatedField(typ reflect.Type, f interface{}) (*graphql.Field, error) {
+	fun := reflect.ValueOf(f)
+
+	ptr := reflect.PtrTo(typ)
+
+	if fun.Kind() != reflect.Func {
+		return nil, fmt.Errorf("fun must be func, not %s", fun)
+	}
+
+	funcType := fun.Type()
+
+	in := make([]reflect.Type, 0, funcType.NumIn())
+	for i := 0; i < funcType.NumIn(); i++ {
+		in = append(in, funcType.In(i))
+	}
+
+	var argParser *argParser
+	var argType graphql.Type
+	var ptrFunc bool
+	var hasContext, hasSource, hasArgs, hasSelectionSet bool
+
+	if len(in) > 0 && in[0] == contextType {
+		hasContext = true
+		in = in[1:]
+	}
+	if len(in) > 0 && (in[0] == typ || in[0] == ptr) {
+		hasSource = true
+		ptrFunc = in[0] == ptr
+		in = in[1:]
+	}
+
+	if len(in) > 0 && in[0] != selectionSetType {
+		hasArgs = true
+		var err error
+		if argParser, argType, err = sb.makeStructParser(in[0]); err != nil {
+			return nil, fmt.Errorf("attempted to parse %s as arguments struct, but failed: %s", in[0].Name(), err.Error())
+		}
+		if argParser, argType, err = sb.hackyArgParser(reflect.TypeOf(ConnectionArgs{}), argParser, argType); err != nil {
+			return nil, fmt.Errorf("attempted to wrap %s as arguments struct, but failed: %s", in[0].Name(), err.Error())
+		}
+		in = in[1:]
+	}
+
+	if len(in) > 0 && in[0] == selectionSetType {
+		hasSelectionSet = true
+		in = in[:len(in)-1]
+	}
+
+	// We have succeeded if no arguments remain.
+	if len(in) != 0 {
+		return nil, fmt.Errorf("%s arguments should be [context][, [*]%s][, args][, selectionSet]", funcType, typ)
+	}
+
+	// Parse return values. The first return value must be the actual value, and
+	// the second value can optionally be an error.
+
+	out := make([]reflect.Type, 0, funcType.NumOut())
+	for i := 0; i < funcType.NumOut(); i++ {
+		out = append(out, funcType.Out(i))
+	}
+
+	var hasRet, hasError bool
+
+	if len(out) > 0 && out[0] != errType {
+		hasRet = true
+		out = out[1:]
+	}
+
+	if len(out) > 0 && out[0] == errType {
+		hasError = true
+		out = out[1:]
+	}
+
+	if len(out) != 0 {
+		return nil, fmt.Errorf("%s return values should [result][, error]", funcType)
+	}
+
+	if !hasRet {
+		return nil, fmt.Errorf("%s has no return value", funcType)
+	}
+
+	var retType graphql.Type
+	if hasRet {
+		var err error
+		retType, err = sb.constructConnType(typ)
+		if err != nil {
+			return nil, err
+		}
+
+	} else {
+		var err error
+		retType, err = sb.getType(reflect.TypeOf(true))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	args := make(map[string]graphql.Type)
+	if hasArgs {
+		inputObject, ok := argType.(*graphql.InputObject)
+		if !ok {
+			return nil, fmt.Errorf("%s's args should be an object", funcType)
+		}
+
+		for name, typ := range inputObject.InputFields {
+			args[name] = typ
+		}
+	}
+
+	ret := &graphql.Field{
```

Comment:
this anon function might seem similar however i dont think this can be refactored